File: code/examples/03_contracts.py
"""
03 — Contract analysis

Extract key clauses and metadata from legal contracts.
Demonstrates callbacks to process each document as it completes.

Requirements:
    pip install sifter-ai
    # Sifter server running on localhost:8000 (./run.sh)
"""
from sifter import Sifter

s = Sifter(api_key="sk-dev")

sift = s.create_sift(
    name="Contracts Q1",
    instructions="""
    Extract the following from each contract:
    - contract_type: e.g. NDA, service agreement, employment, supply
    - party_a: first party (full legal name)
    - party_b: second party (full legal name)
    - effective_date: when the contract takes effect (YYYY-MM-DD)
    - expiry_date: when it expires, if specified (YYYY-MM-DD or null)
    - governing_law: jurisdiction / applicable law
    - termination_notice_days: days notice required to terminate, if stated
    - auto_renewal: true/false — does it renew automatically?
    - key_obligations: brief summary of main obligations (max 2 sentences)
    - penalty_clauses: any penalties or liquidated damages mentioned (or null)
    """,
)

# Track completion
processed = []

def on_document_done(doc_id: str, record: dict):
    data = record.get("extracted_data", record)
    contract_type = data.get("contract_type", "unknown")
    parties = f"{data.get('party_a', '?')} ↔ {data.get('party_b', '?')}"
    print(f"  ✓ [{contract_type}] {parties}")
    processed.append(record)

sift.on("sift.document.processed", on_document_done)
sift.on("sift.completed", lambda sift_id: print(f"\nAll done — {len(processed)} contracts processed."))

sift.upload("./examples/contracts/")
sift.wait()

# Flag contracts expiring in the next 90 days
from datetime import date, timedelta
threshold = date.today() + timedelta(days=90)

expiring_soon = []
for r in sift.records():
    data = r.get("extracted_data", r)
    expiry = data.get("expiry_date")
    if expiry:
        try:
            expiry_date = date.fromisoformat(str(expiry))
            if expiry_date <= threshold:
                expiring_soon.append((data.get("party_a"), data.get("party_b"), expiry_date))
        except ValueError:
            pass

if expiring_soon:
    print("\nContracts expiring within 90 days:")
    for party_a, party_b, expiry in sorted(expiring_soon, key=lambda x: x[2]):
        print(f"  {expiry}  {party_a} ↔ {party_b}")

File: code/examples/ts/03_contracts.ts
/**
 * 03 — Contract analysis
 *
 * Extract key clauses and metadata from legal contracts.
 * Demonstrates callbacks to process each document as it completes.
 *
 * Requirements:
 *   npx tsx 03_contracts.ts
 *   # Sifter server running on localhost:8000
 *   # SIFTER_API_KEY env var set
 */
import { Sifter } from "@sifter-ai/sdk";
import type { SiftRecord } from "@sifter-ai/sdk";

const s = new Sifter();

const sift = await s.createSift(
  "Contracts Q1 TS",
  "These are legal contracts. Extract the contract type, the two parties involved, the effective and expiry dates, the governing law, how many days notice is required to terminate, whether it renews automatically, a brief summary of key obligations and any penalty clauses mentioned.",
);

// Track completion
const processed: unknown[] = [];

sift.on("sift.document.processed", (docId: unknown, record: unknown) => {
  const r = record as SiftRecord;
  const data = r.extracted_data ?? (record as Record<string, unknown>);
  const contractType = (data["contract_type"] as string) ?? "unknown";
  const partyA = (data["party_a"] as string) ?? "?";
  const partyB = (data["party_b"] as string) ?? "?";
  console.log(`  ✓ [${contractType}] ${partyA} ↔ ${partyB}`);
  processed.push(record);
});

sift.on("sift.completed", () => {
  console.log(`\nAll done — ${processed.length} contracts processed.`);
});

await sift.upload("../documents/contracts/", { onConflict: "replace" });
await sift.wait();

// Flag contracts expiring in the next 90 days
const threshold = new Date();
threshold.setDate(threshold.getDate() + 90);

const expiringSoon: Array<{ partyA: string; partyB: string; expiry: Date }> = [];

for await (const r of sift.iterRecords<SiftRecord>()) {
  const data = r.extracted_data ?? (r as unknown as Record<string, unknown>);
  const expiryStr = data["expiry_date"] as string | null;
  if (expiryStr) {
    try {
      const expiryDate = new Date(expiryStr);
      if (!isNaN(expiryDate.getTime()) && expiryDate <= threshold) {
        expiringSoon.push({
          partyA: (data["party_a"] as string) ?? "?",
          partyB: (data["party_b"] as string) ?? "?",
          expiry: expiryDate,
        });
      }
    } catch { /* skip */ }
  }
}

if (expiringSoon.length > 0) {
  console.log("\nContracts expiring within 90 days:");
  expiringSoon.sort((a, b) => a.expiry.getTime() - b.expiry.getTime());
  for (const { partyA, partyB, expiry } of expiringSoon) {
    console.log(`  ${expiry.toISOString().slice(0, 10)}  ${partyA} ↔ ${partyB}`);
  }
}
